### PR TITLE
HCMPRE-169 code review comments

### PR DESCRIPTION
**Pull request by Priyanka-eGov**
## Summary by CodeRabbit

- **New Features**
	- Added a new dependency to improve modularity and capabilities of the health services.
	- Introduced new constants for better error handling related to invalid plan IDs.
	- New method to count `PlanConfiguration` entities based on search criteria.

- **Improvements**
	- Enhanced tenant ID handling in validation logic for better clarity and maintainability.
	- Simplified response generation in `PlanConfigController` for creating plan configurations.

- **Documentation**
	- Improved code documentation and comments for better understanding of methods and their functionalities. 

- **Refactor**
	- Refactored multiple services to utilize modern Java practices, improving code readability and maintainability.

- **Bug Fixes**
	- Fixed potential inconsistencies in error message formatting.

### CODE REVIEW

**Comment by Priyanka-eGov** [comment]
@coderabbitai review

File: health-services/plan-service/src/main/java/digit/service/PlanValidator.java [hunk 1]
```diff
@@ -351,33 +375,51 @@ public void validateTargetMetrics(PlanRequest request, Object mdmsData) {
...
 
-        for (Target target : plan.getTargets()) {
-            if (!metricListFromMDMS.contains(target.getMetric())) {
-                throw new CustomException(METRIC_NOT_FOUND_IN_MDMS_CODE, METRIC_NOT_FOUND_IN_MDMS_MESSAGE);
-            }
-        }
+        List<Object> finalMetricListFromMDMS = metricListFromMDMS;
+        plan.getTargets().stream()
+                .map(Target::getMetric)
+                .filter(metric -> !finalMetricListFromMDMS.contains(metric))
```
Comment (shashwat-egov) [inline health-services/plan-service/src/main/java/digit/service/PlanValidator.java:381]: final metric list needs to be changed to a hash set for validation

File: health-services/plan-service/src/main/java/digit/service/PlanValidator.java [hunk 2]
```diff
@@ -351,33 +375,51 @@ public void validateTargetMetrics(PlanRequest request, Object mdmsData) {
...
         }
 
-        for (MetricDetail metricDetail : metricDetails) {
-            if (!metricUnitListFromMDMS.contains(metricDetail.getMetricUnit())) {
-                throw new CustomException(METRIC_UNIT_NOT_FOUND_IN_MDMS_CODE, METRIC_UNIT_NOT_FOUND_IN_MDMS_MESSAGE);
-            }
-        }
+        metricDetails.stream()
+                .map(MetricDetail::getMetricUnit)
+                .filter(metricUnit -> !metricUnitListFromMDMS.contains(metricUnit))
```
Comment (shashwat-egov) [inline health-services/plan-service/src/main/java/digit/service/PlanValidator.java:417]: change metric list to metric hash set

File: health-services/plan-service/src/main/java/digit/service/enrichment/EnrichmentService.java [hunk 3]
```diff
@@ -64,6 +54,8 @@ public void enrichPlanConfiguration(PlanConfiguration planConfiguration) {
 
         // Generate id for resource mappings
         planConfiguration.getResourceMapping().forEach(resourceMapping -> UUIDEnrichmentUtil.enrichRandomUuid(resourceMapping, "id"));
+
+        enrichAuditDetails(request, Boolean.TRUE);
```
Comment (shashwat-egov) [inline health-services/plan-service/src/main/java/digit/service/enrichment/EnrichmentService.java:58]: refer method from services common library for enrichining audit details

File: health-services/plan-service/src/main/java/digit/service/enrichment/EnrichmentService.java [hunk 4]
```diff
@@ -84,49 +76,38 @@ public void enrichAuditDetails(PlanConfigurationRequest request, Boolean isCreat
...
 
-        // For ResourceMappings
+        // Generate id for ResourceMappings
         planConfiguration.getResourceMapping().forEach(resourceMapping -> {
             if (ObjectUtils.isEmpty(resourceMapping.getId())) {
                 UUIDEnrichmentUtil.enrichRandomUuid(resourceMapping, "id");
             }
         });
 
+        enrichAuditDetails(request, Boolean.FALSE);
```
Comment (shashwat-egov) [inline health-services/plan-service/src/main/java/digit/service/enrichment/EnrichmentService.java:110]: refer servies common audit details enrichment utility

**Review by shashwat-egov — CHANGES_REQUESTED** [review]

File: health-services/plan-service/src/main/java/digit/service/PlanValidator.java [hunk 5]
```diff
@@ -290,42 +281,61 @@ public void validatePlanUpdate(PlanRequest request) {
...
         Set<String> targetUuids = request.getPlan().getTargets().stream()
                 .map(Target::getId)
                 .collect(Collectors.toSet());
 
-        // If target uuids are not unique, throw an exception
-        if(targetUuids.size() != request.getPlan().getTargets().size()) {
-            throw new CustomException("DUPLICATE_TARGET_UUIDS", "Target uuids should be unique");
+        // If target UUIDs are not unique, throw an exception
+        if (targetUuids.size() != request.getPlan().getTargets().size()) {
+            throw new CustomException("DUPLICATE_TARGET_UUIDS", "Target UUIDs should be unique");
```
Comment (shashwat-egov) [inline health-services/plan-service/src/main/java/digit/service/PlanValidator.java:298]: Move these string literals to error codes

File: health-services/plan-service/src/main/java/digit/service/PlanValidator.java [hunk 6]
```diff
@@ -290,42 +281,61 @@ public void validatePlanUpdate(PlanRequest request) {
...
         Set<String> resourceUuids = request.getPlan().getResources().stream()
                 .map(Resource::getId)
                 .collect(Collectors.toSet());
 
-        // If resource uuids are not unique, throw an exception
-        if(resourceUuids.size() != request.getPlan().getResources().size()) {
-            throw new CustomException("DUPLICATE_RESOURCE_UUIDS", "Resource uuids should be unique");
+        // If resource UUIDs are not unique, throw an exception
+        if (resourceUuids.size() != request.getPlan().getResources().size()) {
+            throw new CustomException("DUPLICATE_RESOURCE_UUIDS", "Resource UUIDs should be unique");
```
Comment (shashwat-egov) [inline health-services/plan-service/src/main/java/digit/service/PlanValidator.java:316]: move these string literals to error codes

File: health-services/plan-service/src/main/java/digit/service/PlanValidator.java [hunk 7]
```diff
@@ -290,42 +281,61 @@ public void validatePlanUpdate(PlanRequest request) {
...
         Set<String> activityUuids = request.getPlan().getActivities().stream()
                 .map(Activity::getId)
                 .collect(Collectors.toSet());
 
-        // If activity uuids are not unique, throw an exception
-        if(activityUuids.size() != request.getPlan().getActivities().size()) {
-            throw new CustomException("DUPLICATE_ACTIVITY_UUIDS", "Activity uuids should be unique");
+        // If activity UUIDs are not unique, throw an exception
+        if (activityUuids.size() != request.getPlan().getActivities().size()) {
+            throw new CustomException("DUPLICATE_ACTIVITY_UUIDS", "Activity UUIDs should be unique");
```
Comment (shashwat-egov) [inline health-services/plan-service/src/main/java/digit/service/PlanValidator.java:334]: move this string literal to error codes

File: health-services/plan-service/src/main/java/digit/service/validator/PlanConfigurationValidator.java [hunk 8]
```diff
@@ -380,45 +439,45 @@ public void validateResourceMappingAgainstMDMS(PlanConfigurationRequest request,
...
+     * returns the list of columns which have the property 'isRequired' as true
+     *
+     * @param schemas  List of schemas from MDMS
+     * @param templateIds The list of template identifiers from request object
+     * @param inputFileTypes The list of input file type from request object
+     * @return List of Columns that are required
      */
-    public static List<String> getIsTruePropertyFromSchema(List<Object> schemas, List<String> templateIds, List<String> inputFileTypes) {
+    public static HashSet<String> getRequiredColumnsFromSchema(HashSet<Object> schemas, List<String> templateIds, List<String> inputFileTypes) {
         if (schemas == null) {
```
Comment (shashwat-egov) [inline health-services/plan-service/src/main/java/digit/service/validator/PlanConfigurationValidator.java:451]: use CollectionUtils.isEmpty to perform null and empty checks

**Review by shashwat-egov — CHANGES_REQUESTED** [review]

**Review by shashwat-egov — APPROVED** [review]

**Comment by Priyanka-eGov** [comment]
https://digit-discuss.atlassian.net/browse/HCMPRE-169
This is the ticket @kavi-egov

**Review by kavi-egov — APPROVED** [review]

**Pull request merged**